Add NodeReader interface

mildred opened the pull request:
- [x] jbenet/go-multicodec#9: cbor & json: add HeaderPath variable
- [x] whyrusleeping/cbor#3: Allow streaming interface to CBOR
- [x] implement NodeReader for protocol buffers
- [ ] [Check multicodec headers](#discussion_r54835160) **(and update the spec)**

---

See issue ipfs/go-ipld#17

Do not merge yet, tests are still needed, and I would like to use this interface in some places instead of the Node type. This will allow implementations that do not require buffering the whole node in memory.

mildred commented:
I added tests for NodeReader, and I believe it is ready to be used. I have a few things that I would still like to improve. @jbenet, tell me what you think of it.

- The `Walk` function will never work with `NodeReader`. The callback takes a current `Node` as argument. Is it desirable to design a similar function for NodeReader? If so, how do we avoid buffering data?

- The `Link` type is just a `Node` in desguise. It has many methods that I don't think we need yet, especially considering the IPLD spec ipfs/specs#37. I'd also like to replace it with an interface (and have `Node` implements it).

Once this is done, I'd like to make sure that ipld implements the draft spec (and perhaps drop `@` character escaping as this is not part of the spec). Make sure it works well, and integrate it to the github.com/ipfs/go-ipfs/merkledag package. Mostly, we'd read a `NodeReader` in one pass to build a `merkledag.Node`, dropping non necessary keys.

mildred commented:
Depends on:

- jbenet/go-multicodec#9: cbor & json: add HeaderPath variable
- whyrusleeping/cbor#3: Allow streaming interface to CBOR

It can't compile unless those pull requests are merged (especially the CBOR PR).

jbenet commented:
this is looking like a great start. Im not sure that the interface here should look like. i wont know well until i use it for something... maybe we can make some examples if we want to accelerate that.

mildred commented:
I am not sure the interface should look like this either, but I thought I would take the closest interface I could have to the parsers, with as little overhead as possible. With the idea we could develop other interfaces more convenient for use if necessary.

mildred commented:
I am satisfied with the state of this PR. I think it is ready for review.

Here is how the code is organized:

- what was in go-ipld/ is not in go-ipld/memory/ as this is a memory implementation of a Node
- go-ipld/ is now a package that contain Node interfaces and primitives on these interfaces
- go-ipld/coding/ manages encoding and decoding of a node using CBOR / JSON and Protocol Buffers (decoding only)

How you would want to use it:

- Decoding a node takes a `io.Reader` as argument (`io.ReadSeeker` if you want to read it multiple times, but that is managed using interface upgrades) and returns a `ipld.NodeReader`.
- A NodeReader can be read using the `ipld.NodeIterator` that transform the callback interface of `ipld.NodeReader` into an iterator
- A `ipld.NodeReader` can be copied into a `ipld.NodeWriter` using `ipld.Copy()`. Currently, the only implementation for a NodeWriter is in the memory package
- The coding package takes a `memory.Node` and can write it to CBOR & JSON. There is no support for writing directly from a NodeReader.

The walk functions are in the memory package. I think we need to define what use case we want to have and then implement them for NodeReader in general.

mildred commented:
@jbenet could you look at that, or invite other people who might be interested in looking at this code?

mildred commented:
For example of use, see https://github.com/mildred/go-ipfs/commits/ipld-wip

mildred commented:
Requires whyrusleeping/cbor#4

mildred commented:
Tests are failing for two reasons: the dependency on new features in the CBOR lib, and the dependency to JSON features that are in Go 1.5 but not 1.4 (json token reader)

mildred commented:
For Go 1.4, I vendored the `encoding/json` library that I took from the Go 1.5 release

jbenet commented:
@mildred lmk when you want review here

mildred commented:
There were some issues about the protocol buffer generated file. I regenerated it.

@jbenet I think you can start looking at the code. I will continue to fix the CI in case it is broken. It should not be difficult as the tests are passing on my machine.

mildred commented:
There are many things that were modified. The reader interface has been created and the old `ipld.Node` has been transformed into a `memory.Node`. The codecs were modified so they could decode to a NodeReader and encode from a memory.Node. There is a simple way to convert from a NodeReader to a memory.Node.

Don't hesitate to look at the commits one after another. I rebased them so they follow a natural order. The tests should pass on all commits (or this is a mistake) and a subset of these commits only can be merged if the others are not acceptable.

In particular, I modified how the multicodec format is handled. I found the Multicodec type too impractical to use, mostly because they work on a byte slice and not a stream. Also, the numerous objects, all having an interface type, makes it difficult to know their implementation. The code here still uses the multicodec package to detect the header and encode or decode it, but the Multicodec type is not used.

I followed the IPLD spec about which multicodec headers to use. I am not very sure of these headers. This might need double checking.

IPLD paths are still unimplemented. The old walk algorithm is still there, but only works on a memory.Node.

IPLD link detection is performed on top of the NodeReader interface and is generic to the Link type. it uses struct fields annotations to generate the Link objects. Here `links/link_test.go` might be worth reading.

mildred commented:
The last commit "Regenerate the protocol buffer code" contains a patch to fix the generated code because the field ordering when serializing is wrong. Leading to a hash mismatch in the tests. Upstream issue to remove the patch: gogo/protobuf#151

Kubuxu commented:
Would it be possible to update `_examples` to confront new code?

mildred commented:
What might be nice to add is Marshaller / Unmarshaller functions that does approximately the same as the `links` package but for whole IPLD objects.

Basically:

    // Marshall source into dest
    Marshall(source interface{}, dest memory.Node)

    // Unmarshall the NodeReader to the dest
    Unmarshall(reader NodeReader, dest interface{})

`interface{}` objects could contain ipld annotations like for links.

Kubuxu commented:
It would be awesome, also check if interface{} implements special (Un)Marshalling interface, same way JSON marshaller does it. Thus allowing for custom (Un)Marshalling strategies.

mildred commented:
Here is a first implementation of `Unmarshal()`. It is by no means complete and it will not unmarshal all types. The `Marshal()` function is not there either. This gives out an idea of what is possible.

jbenet commented on coding/cbor/encode.go:
```diff
@@ -0,0 +1,75 @@
+func RawEncode(w io.Writer, node memory.Node, tags bool) error {
+	enc := cbor.NewEncoder(w)
+	if tags {
+		enc.SetFilter(func(val interface{}) interface{} {
```
i believe this closure will generate a function each time. why this and not some simpler, more decomposed thing?

mildred replied:
You're right, this can be avoided
- [x] Separate inline function

jbenet commented on coding/coding.go:
```diff
@@ -1,75 +1,112 @@
+	CodecCBORNoTags
+)
+
+var StreamCodecs map[string]func(io.ReadSeeker) (stream.NodeReader, error)
```
Go is not a functional language. It makes it harder to reason about the code to have `funcs` all over the place. I think these should be an interface, and one that follows the `io` pattern:

``` go
type StreamCodec interface {
  // read first object from stream and populate o
  Read(stream io.ReadSeeker, stream.NodeReader) error

  // write o into stream 
  Write(stream io.ReadSeeker, stream.NodeWriter) error
}
```

and really, why `NodeReader`? it should probably be as simple as just a `ipld.Node` type.

jbenet replied:
btw, this whole module should be using `multicodec` and the coding pattern establish there. i created the `Codec` abstractions for this use case. they achieve something similar.

mildred replied:
There are two things here:
- extensibility is generally better in terms of objects in a OOP language (I would argue with that, especially with a language like Go which allow nice function expressions, but never mind)
- readability. In this case, a single literal expression can be used to declare all codecs. That's really nice and makes it easier to read. Also, it is very easy to know which code is called in a function call. With too much OOP interfaces, it becomes very difficult.

About the multicodec package, I tried first to use it, but it isn't suited to work with streams. See [my last message](#issuecomment-191811920)

jbenet commented on coding/coding_test.go:
```diff
@@ -1,20 +1,22 @@
+	links "github.com/ipfs/go-ipld/links"
+	memory "github.com/ipfs/go-ipld/memory"
+	reader "github.com/ipfs/go-ipld/stream"
+	readertest "github.com/ipfs/go-ipld/stream/test"
```
import as `rt` ? `readertest` everywhere just adds noise

mildred replied:
ok
- [x] rename `readertest` to `rt`

jbenet commented on coding/json.testfile:
```diff
@@ -1,3 +1,3 @@
-/multicodec
+/mdagv1
```
i think this should have a `/multicodec` first still. im not 100% sure, i need to revisit.

mildred replied:
Yes, and also update the IPLD spec accordingly.

jbenet commented on coding/json/stream-go1.4.go:
```diff
@@ -0,0 +1,206 @@
+// +build !go1.5
```
did something change in go1.5?

mildred replied:
Go 1.4 does not have JSON token streaming

mildred replied:
By the way, this file is generated from stream.go. It is almost identical.

jbenet commented on coding/json/stream-go1.4.go:
```diff
@@ -0,0 +1,206 @@
+		return err
+	}
+	//log.Printf("JSON: read token value %#v %T", token, token)
+	if delim, ok := token.(json.Delim); ok {
```
yikes, no. this package should not have any json parsing. if no stream json parser exists that has hooks in it to add our callbacks, either this should become one, or just forget the stream parsing. 

stream parsing json is already a bad idea and if people are doing it they're going to have a bad time anyway.

mildred replied:
The json parsing is actually done in the json package. This is just a token reader. It happens that the tokens just look like json delimiters, but this is no parsing.

See:
- https://godoc.org/encoding/json#example-Decoder-Decode-Stream
- https://godoc.org/encoding/json#example-Decoder-Token

jbenet commented on coding/pb/ipld.pb.go.patch:
```diff
@@ -0,0 +1,28 @@
+--- ipld.pb.go.old	2016-02-24 10:29:39.045261549 +0100
++++ ipld.pb.go.new	2016-02-24 10:29:45.038594961 +0100
```
why is this patch here?

mildred replied:
See gogo/protobuf#151 (the field order is wrong when regenerating the protocol buffer code)

jbenet commented on memory/node.go:
```diff
@@ -187,3 +188,91 @@ func LinkCast(v interface{}) (l Link, ok bool) {
+	return err
+}
+
+func read(curr interface{}, fun stream.ReadFun, path []interface{}) error {
```
im sure this function could be a lot simpler. the repetition shows there's a missing abstraction

mildred replied:
I think this may be difficult. Also, I don't like abstractions just for the sake of them. Sometimes code is easier to read with a little repetition.

But still, this may be needing more thought
- [x] rework `memory.Node.Read`

mildred replied:
Not much can be changed, it's just going through the datastructure. I could improve the NodeReadSkip lines though

jbenet commented on memory/reader_test.go:
```diff
@@ -0,0 +1,64 @@
+	"testing"
+
+	reader "github.com/ipfs/go-ipld/stream"
+	ldtest "github.com/ipfs/go-ipld/stream/test"
```
`ldt`

mildred replied:
ok
- [x] rename `ldtest` to `ldt`

jbenet commented on stream/writer.go:
```diff
@@ -0,0 +1,36 @@
+package stream
+
+type NodeWriter interface {
```
it seems that the entire `stream` package should be under `coding`

mildred replied:
What do you mean? Merge the `stream` and `codec` packages? Perhaps, but this will make a huge package, and the two can live independently. Also, there may be issues with circular package dependencies (mostly in the test files).

jbenet commented on unmarshal.go:
```diff
@@ -0,0 +1,336 @@
+package ipld
```
marshaling and unmarshaling is part of `coding`.

mildred replied:
Not this one.
- In `coding` you have the functions `Encode(...)` and `Decode(...)` that takes a `io.Reader` or `io.Writer`.
- In `stream` you have the `NodeReader` that can be marshalled and unmarshalled into arbitrary datatypes. The Go reflection is used here to perform the trick. It uses type annotation and is independant from the underlying coding (it is generic to the `NodeReader`)

Marshalling and unmarshalling do not transform a fixed binary representation to a structured type (that would be a codec). They take a generic node reader.

mildred replied:
Marshalling and unmarshalling are extra utilities. They shouldn't be in this PR perhaps but included elsewhere.

jbenet commented on unmarshal.go:
```diff
@@ -0,0 +1,336 @@
+	return it.LastError
+}
+
+func Unmarshal(r stream.NodeReader, dest interface{}) error {
```
why do we need this whole thing here? do we need a general `Unmarshal` in this package? why cant we just avoid all the reflection mess by forcing: 

``` go
Unmarshal(in ______, dest ipld.Node) error { ... }
```

mildred replied:
I don't quite understand what you mean with `Unmarshal(in ______, dest ipld.Node) error { ... }`.

The idea is to use reflection to use type annotations and allow applications to use their preferred data type. For example, a directory entry might want to declare its IPLD type this way:

```
type UnixFSDirectory struct {
  Mode os.FileMode `ipld:"key:mode"`
  Files map[string]UnixFSLink `ipld:"key:files"`
}

type UnixFSEntry struct {
  Hash mh.Multihash `ipld:"hash"`
  Name string `ipld:"name"`
}
```

This can effortlessly be mapped to an IPLD object with only one copy performed. The Unmarshal function will read directly from the NodeReader to fill these types. And the NodeReader is directly reading from the io.Reader that might be the network socket directly. No extra copies performed when none is needed.

jbenet commented:
@mildred thanks for the hard work on all this implementation. 

However, i think this is way beyond the parameters of this package. This package should be:
- **type definition for ipld objects + links** (main goal!)
- some convenience functions to manipulate ipld objects
- serialization convenience functions that use multicodec for the heavy lifting 

All the `coding` stuff should be a _very thin_ wrapper around multicodec -- the serialization stuff has become enormous here. If you want to create a `StreamingCodec` as part of the `multicodec` package, maybe there's a case for it there. But here, it really is not the domain of ipld to do all this streaming heavy lifting, let alone json parsing.

Also, the main goal has been superceded. The main goal is a nice definition (inteface + functions) for the ipld objects and links. Those have now been relegated to sub packages. If we're going to have different implementations, that's fine, but we should still have an interface in the root, such that end users typically only need to import the root package.

```
import "ipld"

func MyFunc(obj ipld.Node) { ... }
```

mildred commented:
Thanks for the review. I expected this PR to need more design work anyway. I haven't yet answered your other requests  but I think we should define the direction we are going.

> This package should be:
>
> -    type definition for ipld objects + links (main goal!)
> -    some convenience functions to manipulate ipld objects
> -    serialization convenience functions that use multicodec for the heavy lifting

What do you propose that an IPLD object should be ? I came to the conclusion that it is not possible to come up with one data type that would fit all needs, and that we need an interface type. The interface type should allow a streaming interface in case applications want to use large objects that may not be convenient to fit entirely in memory.

My answer to that question is the `NodeReader` type in the stream package. I'm not entirely satisfied with that because the interface (which is just a function with a callback) doesn't seem to be the best for me. But that's the best I could come up with given the above restrictions (interface and streaming).

You may have noticed that there is no longer a single `ipld.Node` type. However there is :

- A `stream.NodeReader` that provides a raw streaming interface, as close to the actual binary encoding as possible.
- A `memory.Node` which is an implementation of the NodeReader in memory using native Go data structures (maps, ...).

I believe every application will want to use its own data structure. That's why there is the `Unmarshal()` function in the stream package that works like `encoding/json.Unmarshal()`. It takes a NodeReader as argument and fill in the application data structure using Go reflection.
- Not appearing here, but application specific types that can be converted from the `stream.NodeReader` using the `Unmarshal` function.

> All the coding stuff should be a very thin wrapper around multicodec -- the serialization stuff has become enormous here. If you want to create a StreamingCodec as part of the multicodec package, maybe there's a case for it there. But here, it really is not the domain of ipld to do all this streaming heavy lifting, let alone json parsing.

I had to change this part quite a bit. The reason I didn't used the multicodec package to its fullest is mostly because I wanted to be stream oriented. As for JSON, this is just a stream reader, not a parser. See my above comment.

Perhaps we should discuss separately about the multicodec package:

- currently the multicodec doesn't know about stream oriented codecs. We can either just go back to `[]byte` oriented codecs, or teach the multicodec package to do streaming
- we may also want to talk about the best interface for this package. I found the multicodec objects very hard to understand because of the many level of indirection required.

> Also, the main goal has been superceded. The main goal is a nice definition (inteface + functions) for the ipld objects and links. Those have now been relegated to sub packages. If we're going to have different implementations, that's fine, but we should still have an interface in the root, such that end users typically only need to import the root package.

I agree that this is a lot of code. Some of it is entirely optional and not required. Perhaps I should split this in multiple PR, but I think this is not yet the time. We should first define what we want. Another reason for this is that I felt quite alone coding this, and had little feedback doing so. Do you have advices on where I could throw new ideas and get feedback on them?

The pending issues I see and I expect input from you are:

- [ ] Should this interface be stream oriented ? *my answer: yes*
- [ ] What should be the generic IPLD interface ? *stream.NodeReader*
- [ ] How applications can interface with the generic IPLD object ? *by having their own types and unmarshaling the NodeReader to their specific types*
- [ ] make sure the implementation can be fast. No unnecessary overhead.
- [ ] what about the multicodec package (make it stream compatible, see if we can simplify it)

davidar commented:
Yeah, this all seems quite a bit more complicated than I would have expected, given that the data model is basically just JSON with links. There appears to be some premature generalisation happening here. Sorry to be blunt.

whyrusleeping commented on coding/cbor/encode.go:
```diff
@@ -0,0 +1,75 @@
+			var linkPath interface{}
+			var linkObject interface{}
+
+			node, ok = val.(memory.Node)
```
should just use a `:=` here instead of predeclaring all these variables

mildred commented:
> However, i think this is way beyond the parameters of this package. This package should be:
>
> -   type definition for ipld objects + links (main goal!)
> -   some convenience functions to manipulate ipld objects
> -   serialization convenience functions that use multicodec for the heavy lifting

I couldn't sleep and thought of an alternative implementation that seems better to me (perhaps a little less efficient, but much more usable). I'll implement it in the following days (unless you tell me it is not wise to do so) and will split this PR with the new interface into smaller ones.

@jbenet and everyone else, I wouldn't mine a quick and short note
 to tell me if I'm going in the right direction.
The idea was to replace the `NodeReader` that is not very practical to use by an interface that is more object oriented and much more easy to use. Allowing us to simplify IPLD by removing some useless abstractions.

```
package ipld

import (
	"errors"
)

var ErrEnd = errors.New("IpldIterateEnd")

type Node interface{}

type NodeObject interface {
	Reset() error
	Next() (key string, value Node, err error)
}

type partialNodeObjectMemory interface {
	Length() uint64
	Get(key string) (value Node, err error)
}

type NodeObjectMemory interface {
	NodeObject
	partialNodeObjectMemory
}

type NodeObjectLink interface {
	NodeObject
	LinkPath() string // should be multiaddr ???
}

type NodeObjectLinkMemory interface {
	NodeObjectLink
	partialNodeObjectMemory
}

type NodeArray interface {
	Reset() error
	Next() (value Node, err error)
}

type NodeArrayMemory interface {
	NodeArray
	Length() uint64
	Get(idx uint64) (value Node, err error)
}

func ToObject (n Node) NodeObject { ... } // return nil if it is not an object
func ToObjectMemory (n Node) NodeObjectMemory { ... } // iota
func ToArray (n Node) NodeArray { ... }
func ToArrayMemory (n Node) NodeArrayMemory { ... }
func ToUint (n Node) *uint64 { ... }
func ToInt (n Node) *int64 { ... }
func ToFloat (n Node) *float64 { ... }
```

`ipld/codecs` would take or return a `Node` object. Most of the other packages that are present in this PR would be additional tools not required for normal use (contrary to the present where most are necessary for round tripping).

Also, this interface will allow making use of the special CBOR encoding of links that the `NodeReader` interface is not able to use fully.

One thing I would probably not change is that I will not use the multicodec package fully (by implementing the Multicodec type) because the Multicodec type is buffer oriented, not stream oriented, and because I find it too complex. Unless we can work this out somehow.

mildred commented:
The interface above was almost impossible to implement. I found a better interface that works (I got a first implementation):

```
// Represents a node object that is iterable.
// It iterates over key value pairs. When created, the NodeIterator does not
// have a current item and iteration must be started using Next()
type NodeIterator interface {
	// Go to next item. Sub items iteration is aborted.
	// Returns true if there is an item
	// Returns false at the end of the iteration
	Next() bool

	// Iterate the current item value
	// Return nil if the iteration hasn't started, is finished or if the current
	// item is not iterable
	Iterate() NodeIterator

	// Abort iteration. Does nothing if the iteration is finished. Also aborts
	// child iterators that were returned by the Iterate() method.
	Abort()

	// Return the current item key, or nil if the iteration hasn't started or is
	// finished
	Key() interface{}

	// Return the current item value or nil if the iteration hasn't started, is
	// finished, or if the current item is iterable. In that case the Iterate()
	// method should be called instead.
	Value() interface{}
}
```

mildred commented:
@jbenet @davidar, please review these commits. Don't hesitate to look at the code commit by commit. The commits are mostly independent and I rebased the whole thing so each commit stands on its own.

Some features from last time are not included (they are still in the `reader-extra` branch on my repo). There is only the code features needed.

The basic interface in in `go-ipld/node.go`, it is `NodeIterator` and is easy to work with. bytes can be decoded into it with zero copy and zero buffering. Encoding it back requires to store the whole thing into the memory for a few reasons. Overall, this is not a problem because we will be reading nodes much more often than writing them.

I'm still not using the `Multicodec` type because it is not stream oriented, and also because I think it add too much overhead for something really simple (take a reader, spit out an empty interface). No need for 3 kind of objects per multicodec layer. @jbenet, if you want to make use of the Multicodec, we'll have to figure out how to do streaming with it.

Also, implementing multicodecs without the `Multicodec` type (the multicodec package is still used to read / write headers) is dead simple. Encoding is just a `Write(Header)` and decoding is just a `ReadHeader()` followed by a switch statement (or equivalent)

mildred commented:
(some tests are failing, this is a data race, I'll fix it, but that should not impact the review very much)

davidar commented:
@mildred I can't comment on golang specifics ~~other than "oh god so much `if err != nil` --- where are my monads?!" :p~~ but it seems a lot more focused than before, so LGTM :)

jbenet commented:
I still havent reviewed the code sorry.

Up front, i think the iterator route is great! much, much nicer to do that.

I would go for:

```go
type NodeIterator interface {
  Next() bool // in streaming implementations, automatically Skips all children iterators
  Children() NodeIterator  // instead of Iterate
  Skip() // instead of abort.

  Key() interface{}
  Value() interface{}
}
```

jbenet commented:
- why is the `Key()` an `interface{}`? and not a `string`?

jbenet commented:
With code like:
```go
iter := iterateOverReader(reader)
printLinks(iter)


def printLinks(iter NodeIterator) {
  for iter.Next() {
    if c.Key() == "/" {
      fmt.Println("found link to", c.Value())
    }

    printLinks(iter.Children())
  }
}
```

mildred commented:
On Sat, 21 May 2016 22:26:51 -0700
Juan Benet <notifications@github.com> wrote:

> - why is the `Key()` an `interface{}`? and not a `string`?

The key can be an integer (for arrays) or a string (for maps/objects).

Perhaps that  could be separated in two functions (Key() and Index()). That should be easy, but what should Key() return on objects?

-- 
Mildred Ki'Lya <mildred@mildred.fr>

mildred commented:
Until now I didn't see you wanted renames in the NodeInterface. I performed these.

@jbenet, could you look into this PR, or tell me what I can do to help the review (splitting, ...). Every commit is understandable by itself if that can help. I take care to rebase my changes every time.

Also, this PR is not up to date to the latest spec. In particular the link key is not yet "/" and the code still supports having attributes along with the link itself. I'd like to have it merged before updating these. These modifications are also pending on ipfs/specs#111. I'd like to implement it using proper EJSON.

yusefnapora commented:
Hi @mildred, I've been playing with this branch a bit to try to see how I can help with go-ipld, since we're planning to use it in the mediachain project.  I'm still learning my way around go, so I'll probably miss some obvious things :)

I noticed that links aren't getting encoded to cbor tags, which seems to be because there's two `LinkTags` defined.  The memory nodes use `"mlink"` as the `LinkKey` (defined in `memory/node.go`), but the cbor encoder's `encodeFilter` is checking for `ipld.LinkKey`, which is defined as `@link`.  I'm guessing those are both eventually going to be changed to `/` to fit the new spec?   At the moment the result is links get encoded to cbor as maps with `"mlink"` keys instead of `CBORTag`s.  That matches what's in `cbor.testfile`, so the tests all pass.

Also the `Makefile` in the coding dir references a now-missing file `bin/coding.go`, so I can't easily regenerate the test cases.  Not a big deal; as I was mucking about I generated some manual test cases without too much fuss by just spitting out some cbor and prepending the multicodec header from the existing case.  

Overall this seems great, although I don't really know enough go for that opinion to have much weight 😄 I really like the `NodeIterator` interface though; it seems like it provides a lot of flexibility at the application level.

Anyway, if there's anything I can help you with, please let me know.
